[WebGPUSwift] Swift Implementation of beginComputePass

nmahendru opened the pull request:
#### 5b926e148a538c01693807f22e001b679981504e
<pre>
[WebGPUSwift] Swift Implementation of beginComputePass
<a href="https://rdar.apple.com/141069436">rdar://141069436</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=284195">https://bugs.webkit.org/show_bug.cgi?id=284195</a>

Reviewed by Mike Wyrzykowski.

It&apos;s 1-to-1 replacement.
In some of the failure cases I have added extra returns for
returning an invalid ComputePassEncoder. They only show up for cases
where swift maps the Cpp types as optionals.

* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(CommandEncoder_beginComputePass_thunk(_:descriptor:)):
(WebGPU.beginComputePass(_:)):
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
(refComputePassEncoder):
(derefComputePassEncoder):
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:

Canonical link: <a href="https://commits.webkit.org/288419@main">https://commits.webkit.org/288419@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb292cc6ba8e4e93d71a480c449a0ef7766348cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83189 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/2821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/37482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/88287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/34223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/2900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/10754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/88287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/34223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86243 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/2900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/37482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/88287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/2900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/37482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/33261 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/2900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/37482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/89656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/10467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/10754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/89656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/10696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/37482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/89656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/37482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12859 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/10421 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/15908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/10285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/13753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/12054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -930,4 +935,53 @@ extension WebGPU.CommandEncoder {
+        }
+        let computePassDescriptor = MTLComputePassDescriptor()
+        computePassDescriptor.dispatchType = MTLDispatchType.serial
+        var counterSampleBuffer: Optional<MTLCounterSampleBuffer> = nil
```
```suggestion
        var counterSampleBuffer: MTLCounterSampleBuffer? = nil
```

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -930,4 +935,53 @@ extension WebGPU.CommandEncoder {
+            counterSampleBuffer = WebGPU.fromAPI(wgpuTimestampWrites.pointee.querySet).counterSampleBuffer()
+        }
+
+        if (m_device.ptr().enableEncoderTimestamps() || counterSampleBuffer != nil) {
```
no parentheses

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -930,4 +935,53 @@ extension WebGPU.CommandEncoder {
+
+        self.setExistingEncoder(computeCommandEncoder)
+
+        computeCommandEncoder.label = Swift.String(cString: descriptor.label)
```
`Swift.` isn't needed I don't think

rr-codes commented on Source/WebGPU/WebGPU/ComputePassEncoder.h:
```diff
@@ -106,6 +109,17 @@ class ComputePassEncoder : public WGPUComputePassEncoderImpl, public RefCounted<
     NSString *m_lastErrorString { nil };
     bool m_passEnded { false };
-};
+}SWIFT_SHARED_REFERENCE(refComputePassEncoder, derefComputePassEncoder);
```
missing space before SWIFT_

mwyrzykowski commented on LayoutTests/platform/mac-wk2/TestExpectations:
```diff
@@ -1592,7 +1592,7 @@ webkit.org/b/261356 [ Debug ] fast/mediastream/device-change-event-2.html [ Pass
 
 # WebGPU
-[ Release ] http/tests/webgpu/webgpu/api/validation/encoding/cmds/copyTextureToTexture.html
+[ Release ] http/tests/webgpu/webgpu/api/validation/encoding/cmds
```
I think this should say:
```
[ Release ] http/tests/webgpu/webgpu/api/validation/encoding/cmds [ Pass ]
```
at least locally I need to add `[ Pass ]` to avoid the warning about test passed but expected to fail

mwyrzykowski commented on LayoutTests/platform/mac-wk2/TestExpectations:
```diff
@@ -1595,7 +1595,8 @@ webkit.org/b/261356 [ Debug ] fast/mediastream/device-change-event-2.html [ Pass
 # WebGPU
-[ Release ] http/tests/webgpu/webgpu/api/validation/encoding/cmds/copyTextureToTexture.html
+[ Release ] http/tests/webgpu/webgpu/api/validation/encoding/cmds [ Pass ]
+[ Release ] http/tests/webgpu/webgpu/api/validation/encoding/cmds/render/draw.html [ Skip ]
```
Maybe we can't enable these for now, we see:
```
  http/tests/webgpu/webgpu/api/validation/encoding/cmds/copyTextureToTexture.html [ Pass Failure ]
  http/tests/webgpu/webgpu/api/validation/encoding/cmds/render/dynamic_state.html [ Pass Failure ]
  http/tests/webgpu/webgpu/api/validation/encoding/cmds/render/indirect_draw.html [ Pass Failure ]
  http/tests/webgpu/webgpu/api/validation/encoding/cmds/render/setVertexBuffer.html [ Pass Failure ]
  http/tests/webgpu/webgpu/api/validation/encoding/cmds/render/state_tracking.html [ Pass Failure ]
```
which is unrelated to your source changes

nmahendru committed:
[WebGPUSwift] Swift Implementation of beginComputePass
rdar://141069436
https://bugs.webkit.org/show_bug.cgi?id=284195

Reviewed by Mike Wyrzykowski.

It's 1-to-1 replacement.
In some of the failure cases I have added extra returns for
returning an invalid ComputePassEncoder. They only show up for cases
where swift maps the Cpp types as optionals.

* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(CommandEncoder_beginComputePass_thunk(_:descriptor:)):
(WebGPU.beginComputePass(_:)):
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
(refComputePassEncoder):
(derefComputePassEncoder):
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:

Canonical link: https://commits.webkit.org/288419@main
